### DISCUSSION TITLE
Differences in behavior between simulation and experiment

### DISCUSSION THREAD

**Discussion by Logiant** [USC-ACTLab/crazyswarm#645] [Q&A]
Hi All,

I have been using the most recent build of crazyswarm to send velocity commands to a handful of crazyflies, and I noticed a couple of significant difference between using the --sim argument vs running an experiment.

1. In simulation I can call yaw() on the crazyflie objects (similar to position()) to get to the current yaw angle. This causes a crash in the experiment, with an error that no function yaw() exists. Is there an alternative way to get the current yaw angle? I haven't found one in the docs.
2. Calling the land function causes the crazyflies to fly straight down when using the --sim argument, but in experiments the crazyflies seem to move back toward their initial positions on the ground instead. I see that the land function is not supported after calling CmdVelocityWorld, so it might be worthwhile to print an error (or just have it do nothing) so that the behaviors match.

I'm using a VICON mocap system, and everything seems to be running correctly aside from these couple of issues. I occasionally get hover_swarm to crash with an "unknown transform" error, but I can post a separate issue when I figure out how to reproduce it.

Thanks,
Logan

**Comment by jpreiss** [comment]
`yaw()` could be implemented for real robots following the pattern of `position()`: https://github.com/USC-ACTLab/crazyswarm/blob/967723df9dc1e27bae1c2725d46a182c8f3c2f80/ros_ws/src/crazyswarm/scripts/pycrazyswarm/crazyflie.py#L337-L339

We would need to decide what to do when using `libobjecttracker` in single-marker mode. Recall that there is no way to estimate the yaw on the PC in single-marker mode, it can only be estimated onboard by sensor fusion with the IMU.

Currently `libobjecttracker` returns an identity rotation when in single-marker mode, and the Crazyswarm server publishes this to ROS directly. IMO, it would be better for `yaw()` to return NaN or throw an exception rather than always report a yaw of zero. This would require changes to the Crazyswarm server and/or `libobjecttracker`.

**Comment by jpreiss** [comment]
for landing, did you call `notifySetpointsStop` in between `cmdVelocityWorld` and `land`?

**Reply by Logiant** [reply]
When landing I did not call notifySetpointsStop. I sent each cf a cmdVelocityWorld of all zeros, then called the land function after one second of sleeping.

**Reply by whoenig** [reply]
The behavior you want in 2) should be supported in both real robots and the simulator, if you use `notifySetpointsStop` for switching from the low-level mode back to the high-level mode.
